discovery_engine/lib/src/domain/assets/data_provider.dart

```diff
@@ -29,10 +31,22 @@ abstract class SetupData with EquatableMixin {
   Object get kpeModel;
   Object get kpeCnn;
   Object get kpeClassifier;
+  Object get availableSources;
 
   @override
-  List<Object?> get props =>
-      [smbertVocab, smbertModel, kpeVocab, kpeModel, kpeCnn, kpeClassifier];
+  List<Object?> get props => [
+        smbertVocab,
+        smbertModel,
+        kpeVocab,
+        kpeModel,
+        kpeCnn,
+        kpeClassifier,
+        availableSources,
+      ];
+
+  Future<AvailableSources> getAvailableSources() {
```

Comment:
:art: I think this should be an abstract method. `SetupData` is already `abstract`.

Reply:
i'm not sure what you mean? the method is not abstract because it has a body, it throws an `UnsupportedError` like in our other abstract classes. also adding an `abstract` keyword to a method is illegal syntax in dart:
```
Members of classes can't be declared to be 'abstract'.
Try removing the 'abstract' keyword. You can add the 'abstract' keyword before the class declaration.
dart(abstract_class_member)
```

Reply:
```suggestion
  Future<AvailableSources> getAvailableSources();
```
(and removing the next lines of the body)

There is little reasons to implement a method which is meant to always be override in all concrete instances. If we do so in other cases we should change it there, too.  Throwing `UnsupportedError` is only for cases where a method is implemented in some of the (non concrete) sub-classes but not in others. E.g. in out case with not implementing the
method we lint/type-check that it will be overridden. By giving it a body like that we do lose the type checks while gaining nothing (as far as I can tell) and having more code.

IMHO `UnsupportedError` is most times an anti-pattern through it sometimes can be use-full to avoid increases of complexity.

Reply:
we can change that, but we should do it in a separate pr for all occurences throughout our code base. for the current implementation i would prefer consistency of the local improvement.